### DISCUSSION TITLE
Coerced Enums

### DISCUSSION THREAD

**Discussion by dbachrach** [colinhacks/zod#730] [Show and tell]
We're using zod with Prisma and Blitz. We were wanting our queries/mutations to be able to accept enums from our prisma schema as well as other enums we have defined. Some of these enums come from the url query parameters or are user-modifiable, so we didn't want to be case sensitive. We've been using this coerced enum helper and thought it might be useful to others:

```typescript
type EnumLike = {
  [k: string]: string | number
  [nu: number]: string
}
export const zCoercedEnum = <T extends EnumLike>(e: T) =>
  z.preprocess((val) => {
    const target = String(val)?.toLowerCase()
    for (const k of Object.values(e)) {
      if (String(k)?.toLowerCase() === target) {
        return k
      }
    }

    return null
  }, z.nativeEnum(e))
```

**Comment by matthova** [comment]
Just wrote something very similar. It may be a nice addon for zod to export its `EnumLike` definition from the library. Was thinking this could be a cool addition to the api too, but on second thought, it'd be tough to push as an API since the cool part is doing string case manipulation (or whatever custom thing someone wants to do). I wonder if it'd be worth adding an example like`zod.coercedEnum` directly to the library docs.. Thanks for sharing this snippet!

**Comment by codybrouwers** [comment]
Thanks for the idea! I was trying to find a way to mark a field as an enum but if it's not in the list of values, default it to a specific value and came up with this based on your code:

```tsx
function zodEnumWithDefault<Value extends string, EnumsArray extends Readonly<[Value, ...Value[]]>>(
  values: EnumsArray,
  defaultValue: EnumsArray[number]
) {
  return z.preprocess((incomingValue) => {
    if (typeof incomingValue === "string") {
      if (values.includes(incomingValue as Value)) {
        return incomingValue;
      }
      return defaultValue;
    }
    return null;
  }, z.enum(values));
}
```
